Improve js sdk doc (#244)
directus/docs: reference/sdk.md

> The JS SDK provides an intuitive interface for the Directus API from within a JavaScript-powered project (browsers and
> Node.js). The default implementation uses [Axios](https://npmjs.com/axios) for transport and `localStorage` for
> storing state. Advanced customizations are available.

## Installation

```bash
npm install @directus/sdk
```

## Basic Usage

This is the starting point to use the JS SDK. After you've created the `Directus` instance, you can start invoking
methods from it to access your project and data.

```js
import { Directus } from '@directus/sdk';

const directus = new Directus('http://directus.example.com');
```

### Public Access

You can always access data available to the [public role](/configuration/users-roles-permissions.html#directus-roles).

```js
async function publicData() {
	// We don't need to authenticate if the public role has access.

> 	const publicData = await directus.items('some_public_collection').readByQuery({ meta: 'total_count' });

Reviewer: Lets not focus too heavily on `meta`. Ideally people rely on aggregation rather than the old `meta` query param.

> async function publicData() {
> 	// We don't need to authenticate if the public role has access.
> 	const publicData = await directus.items('some_public_collection').readByQuery({ meta: 'total_count' });
> 
> 	console.log({
> 		items: publicData.data,
> 		total: publicData.meta.total_count,
> 	});
> }

Reviewer: ```suggestion
async function publicData() {
	// We don't need to authenticate if the public role has access.
	const publicData = await directus.items('some_public_collection').readByQuery({ sort: ['id'] });

	console.log(publicData.data);
}
```

			});
	}

	// GET DATA

	// After authentication, we can fetch data that the user has access to.

> 	const privateData = await directus.items('some_private_collection').readByQuery({ meta: 'total_count' });

Reviewer: Same meta thing

start();
```

## Custom Configuration

The previous section covered basic installation and usage of the JS SDK with default configurations for `init`. But
sometimes you may need to customize these defaults.

### Constructor

```js
import { Directus } from '@directus/sdk';

const directus = new Directus(url, init);
```

### Parameters

<br />

#### `url` _required_

- **Type** — `String`
- **Description** — A string that points to your Directus instance. E.g., `https://example.directus.io`
- **Default** — N/A

<br />

#### `init` _optional_

- **Type** — `Object`
- **Description** — Defines authentication, storage and transport settings.
- **Default** — Directus will use the following default values for `init`.

```js
// This is the default init object

> export default {

Reviewer: Lets remove the `export default`s from these, that doesn't really fit in this context (you're not exporting any config, this is the config option you'd pass to Directus). Alternatively, if we don't want to start with just an object floating in space, we could make it:

```suggestion
const config = {
```

	auth: {

> 		mode: typeof window === 'undefined'? 'json': 'cookie',

Reviewer: This is a tad confusing in context; lets go with browser user as the assumed default, and adding a comment for the Node.js environment difference:

```suggestion
		mode: 'cookie', // 'json' in Node.js
```

		autoRefresh: true,
		msRefreshBeforeExpires: 30000,

> 		staticToken: '',

Reviewer: Is this an empty string or `null`?

	},
	storage: {
		prefix: '',
		mode: typeof window == 'undefined' ?  'MemoryStorage': 'LocalStorage'
	},
	transport: {
		params: {},
		headers: {},

> 		onUploadProgress: (ProgressEvent) => void,

Reviewer: ```suggestion
		onUploadProgress: (ProgressEvent) => {},
```

Maybe we can rely on destructuring to hint at what's available in the progress event, like:

```suggestion
		onUploadProgress: ({ percentage }) => void,
```

(`percentage` just an example, I don't know what's available in that event out of the top of my head)

  using multiple instances of SDK.
- **Default** — Defaults to `''` (no prefix).

<br />

#### `mode`

- **Type** — `String`
- **Description** — Defines the storage location to be used to save tokens. Allowed values are `LocalStorage` and
  `MemoryStorage`. The mode `LocalStorage` is not compatible with Node.js. `MemoryStorage` is not persistent, so once
  you leave the tab or quit the process, you will need to authenticate again.
- **Default** — Defaults to `LocalStorage` on browsers and `MemoryStorage` on Node.js.

### Extend `storage`

It is possible to provide a custom implementation by extending `BaseStorage`. While this could be useful in certain
advanced situations, it is not needed for most use-cases.

```js
import { BaseStorage, Directus } from '@directus/sdk';

class SessionStorage extends BaseStorage {
	get(key) {
		return sessionStorage.getItem(key);
	}
	set(key, value) {
		return sessionStorage.setItem(key, value);
	}
	delete(key) {
		return sessionStorage.removeItem(key);
	}
}

const directus = new Directus('https://example.directus.app', {
	storage: new SessionStorage(),
});
```

> ## `transport`

Reviewer 2: ```suggestion
### Transport options
```

});
```

## `storage`

The storage is used to load and save token information. By default, Directus creates an instance of `storage` which
handles store information automatically.

```js
export default {
	prefix: '',
	mode: typeof window === 'undefined' ? 'MemoryStorage' : 'LocalStorage',
};
```

:::tip Multiple Instances

If you want to use multiple instances of the SDK you should set a different [`prefix`](#prefix) for each one.

:::

::: tip

The axios instance can be used for custom requests by calling:

```ts
await directus.transport.<method>('/path/to/endpoint', {
	/* body, params, ... */

> });

Reviewer: I _think_ we might wanna omit this with the eye on the bigger drop-axios sprint that's in the works, at which point this no longer exists.

```

:::

### Options

> <br />

Reviewer: Are these necessary? If we need additional margin/padding, that should probably be solved in CSS rather than with empty lines without meaning

});
```

## TypeScript

Version >= 4.1

> Although it's not required, it is recommended to use Typescript to have an easy development experience. This allows more

Reviewer: ```suggestion
Although it's not required, it is recommended to use TypeScript to have an easy development experience. This allows more
```

detailed IDE suggestions for return types, sorting, filtering, etc.

To feed the SDK with your current schema, you need to pass it on the constructor.

```ts
type BlogPost = {
	id: ID;
	title: string;
};

type BlogSettings = {
	display_promotions: boolean;
};

type MyCollections = {
	posts: BlogPost;
	settings: BlogSettings;
};

// This is how you feed custom type information to Directus.

> const directus = new Directus<MyCollections>('http://url');

Reviewer: Lets use a more "real" demo URL, like

```suggestion
const directus = new Directus<MyCollections>('https://example.com');
```

```

You can also extend the Directus system type information by providing type information for system collections as well.

```ts
import { Directus } from '@directus/sdk';

// Custom fields added to Directus user collection.
type UserType = {
	level: number;
	experience: number;
};

type CustomTypes = {
	/*
	This type will be merged with Directus user type.
	It's important that the naming matches a directus
	collection name exactly. Typos won't get caught here
	since SDK will assume it's a custom user collection.
	*/
	directus_users: UserType;
};

const directus = new Directus<CustomTypes>('https://api.example.com');

await directus.auth.login({
	email: 'admin@example.com',
	password: 'password',
});

const me = await directus.users.me.read();
// typeof me = partial DirectusUser & UserType;

// OK
me.level = 42;

// Error TS2322: Type "string" is not assignable to type "number".
me.experience = 'high';
```

## Authentication

### Get current token

```ts
const token = await directus.auth.token;
```

::: warning Async

Reading the token is an asynchronous getter. This makes sure that any currently active `refresh` calls are being awaited
before the current token is returned.

> :::

Reviewer: Unrelated to this PR, but @freekrai lets see if the updates to SDK is a good time to make this a little less of a footgun. Async getters are not really common practice in JS anno 2022, so it's incredibly easy to make this mistake

Reviewer 3: Yeah, this one has caused some pains before so can see what we can do there.

/* index.js */
import { Directus } from 'https://unpkg.com/@directus/sdk@latest/dist/sdk.esm.min.js';

> const directus = new Directus('https://example.directus.app', {

Reviewer 2: Shouldnt we promote the `// await directus.auth.login({ email, password })` method as the main way of authenticating with the use of static token as advanced sidenote
